File: 94. Binary Tree Inorder Traversal/Solution78.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution78 {
    public static TreeNode createTree(List<Object> nodes) {
        if (nodes == null || nodes.isEmpty()) {
            return null;
        }

        TreeNode root = new TreeNode((Integer) nodes.get(0));
        List<TreeNode> queue = new ArrayList<>();
        queue.add(root);
        int i = 1;

        while (!queue.isEmpty()) {
            TreeNode node = queue.remove(0);
            if (node != null) {
                if (i < nodes.size() && nodes.get(i) != null) {
                    node.left = new TreeNode((Integer) nodes.get(i));
                    queue.add(node.left);
                }
                i++;
                if (i < nodes.size() && nodes.get(i) != null) {
                    node.right = new TreeNode((Integer) nodes.get(i));
                    queue.add(node.right);
                }
                i++;
            }
        }
        return root;
    }

    public static void entrada(List<Object> nodes) {
        TreeNode root = createTree(nodes);
        Solution78 solution = new Solution78();
        System.out.println("Input: root = " + nodes);
        System.out.println("Output: " + solution.inorderTraversal(root));
    }

    public static void main(String[] args) {
        List<Object> nodes = new ArrayList<>();
        nodes.add(1);
        nodes.add(null);
        nodes.add(2);
        nodes.add(3);
        entrada(nodes);
        System.out.println();

        nodes.clear();
        entrada(nodes);
        System.out.println();

        nodes.add(1);
        entrada(nodes);
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> st = new Stack<>();

        while (root != null || !st.isEmpty()) {
            while (root != null) {
                st.push(root);
                root = root.left;
            }

            root = st.pop();
            res.add(root.val);

            root = root.right;
        }

        return res;
    }
}

File: 94. Binary Tree Inorder Traversal/Solution78.py
from typing import Optional, List, Union
import json

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        st = []
        res = []

        while root or st:
            while root:
                st.append(root)
                root = root.left
            
            root = st.pop()
            res.append(root.val)

            root = root.right
        
        return res  

def create_tree(nodes: List[Union[int, None]]) -> Optional[TreeNode]:
    if not nodes:
        return None
    for i in range(len(nodes)):
        if nodes[i] == "null":
            nodes[i] = None
    root = TreeNode(nodes[0])
    queue = [root]
    i = 1
    while queue:
        node = queue.pop(0)
        if node:
            if i < len(nodes) and nodes[i] is not None:
                node.left = TreeNode(nodes[i])
                queue.append(node.left)
            i += 1
            if i < len(nodes) and nodes[i] is not None:
                node.right = TreeNode(nodes[i])
                queue.append(node.right)
            i += 1
    return root

def entrada(nodes): 
    root = create_tree(nodes)
    solution = Solution()
    print("Input: root = ", nodes)
    print("Output:", solution.inorderTraversal(root))

def main():
    nodes = [1, "null", 2, 3] 
    entrada(nodes)
    print()

    nodes = [] 
    entrada(nodes)
    print()

    nodes = [1] 
    entrada(nodes)

   
if __name__ == "__main__":
    main()
